File: articles/calculus-for-systems-modeling/articles/vector-valued-functions-and-motion/julia/trajectory_audit.jl
position(t) = (t, sin(t))
distance_between(p, q) = sqrt((q[1]-p[1])^2 + (q[2]-p[2])^2)

function audit_trajectory(step, scenario)
    times = collect(0.0:step:2.0*pi)
    points = [position(t) for t in times]
    segment_lengths = [distance_between(points[i], points[i+1]) for i in 1:(length(points)-1)]
    speeds = [segment_lengths[i] / (times[i+1]-times[i]) for i in 1:length(segment_lengths)]
    arc_length = sum(segment_lengths)
    displacement = distance_between(points[1], points[end])
    efficiency = displacement / max(arc_length, 1.0e-12)
    warning = step > 0.5 ? "Time step is coarse; turns and speed variation may be undersampled." : "Synthetic trajectory audit; document units, parameter meaning, and sampling."
    return scenario, step, length(points), arc_length, displacement, efficiency, sum(speeds)/length(speeds), maximum(speeds), "trajectory r(t) = <t, sin(t)> for 0 <= t <= 2pi", warning
end

println("scenario,time_step,point_count,approximate_arc_length,displacement_magnitude,path_efficiency,average_speed,maximum_speed,domain_description,warning")
for case in [(1.0,"coarse_time_step"),(0.5,"medium_time_step"),(0.25,"fine_time_step")]
    println(join(audit_trajectory(case[1], case[2]), ","))
end

File: articles/calculus-for-systems-modeling/articles/vector-valued-functions-and-motion/r/trajectory_audit.R
position <- function(t) {
  c(t, sin(t))
}

distance_between <- function(p, q) {
  sqrt((q[1] - p[1])^2 + (q[2] - p[2])^2)
}

audit_trajectory <- function(step, scenario) {
  times <- seq(0, 2 * pi, by = step)
  points <- lapply(times, position)
  segment_lengths <- c()
  speeds <- c()

  for (i in seq_len(length(points) - 1)) {
    segment_length <- distance_between(points[[i]], points[[i + 1]])
    segment_lengths <- c(segment_lengths, segment_length)
    speeds <- c(speeds, segment_length / (times[i + 1] - times[i]))
  }

  arc_length <- sum(segment_lengths)
  displacement <- distance_between(points[[1]], points[[length(points)]])
  efficiency <- displacement / max(arc_length, 1e-12)

  warning <- ifelse(
    step > 0.5,
    "Time step is coarse; turns and speed variation may be undersampled.",
    "Synthetic trajectory audit; document units, parameter meaning, and sampling."
  )

  data.frame(
    scenario = scenario,
    time_step = step,
    point_count = length(points),
    approximate_arc_length = arc_length,
    displacement_magnitude = displacement,
    path_efficiency = efficiency,
    average_speed = mean(speeds),
    maximum_speed = max(speeds),
    domain_description = "trajectory r(t) = <t, sin(t)> for 0 <= t <= 2pi",
    warning = warning
  )
}

results <- rbind(
  audit_trajectory(1.0, "coarse_time_step"),
  audit_trajectory(0.5, "medium_time_step"),
  audit_trajectory(0.25, "fine_time_step")
)

dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_trajectory_audit.csv", row.names = FALSE)
print(results)
